File: glut_opengl_estudos/09_picking_colisoes_aabb.py
"""Aula 09 — Picking por mouse + colisão AABB em tempo real.

Exemplo avançado de engenharia de gameplay 2D com OpenGL clássico:
- Entidades atualizadas com passo temporal fixo por timer.
- Picking de entidade via clique do mouse (NDC).
- Detecção de colisão AABB e realce visual.

Controles:
- Clique esquerdo: seleciona caixa
- R: reset das posições
- ESC: sair
"""

from __future__ import annotations

from dataclasses import dataclass
from random import Random

from OpenGL.GL import (
    GL_COLOR_BUFFER_BIT,
    GL_LINE_LOOP,
    GL_MODELVIEW,
    GL_QUADS,
    glBegin,
    glClear,
    glClearColor,
    glColor3f,
    glEnd,
    glLoadIdentity,
    glMatrixMode,
    glRasterPos2f,
    glVertex2f,
)
from OpenGL.GLUT import (
    GLUT_ACTION_GLUTMAINLOOP_RETURNS,
    GLUT_ACTION_ON_WINDOW_CLOSE,
    GLUT_BITMAP_HELVETICA_12,
    GLUT_DOUBLE,
    GLUT_DOWN,
    GLUT_LEFT_BUTTON,
    GLUT_RGB,
    GLUT_WINDOW_HEIGHT,
    GLUT_WINDOW_WIDTH,
    glutBitmapCharacter,
    glutCreateWindow,
    glutDisplayFunc,
    glutGet,
    glutInit,
    glutInitDisplayMode,
    glutInitWindowSize,
    glutKeyboardFunc,
    glutMainLoop,
    glutMouseFunc,
    glutPostRedisplay,
    glutSetOption,
    glutSwapBuffers,
    glutTimerFunc,
)


@dataclass
class BoxEntity:
    """Representa uma caixa 2D com velocidade, cor e estado de seleção."""

    x: float
    y: float
    half_size: float
    vx: float
    vy: float
    color: tuple[float, float, float]
    selected: bool = False

    def min_x(self) -> float:
        return self.x - self.half_size

    def max_x(self) -> float:
        return self.x + self.half_size

    def min_y(self) -> float:
        return self.y - self.half_size

    def max_y(self) -> float:
        return self.y + self.half_size

    def contains(self, px: float, py: float) -> bool:
        """Retorna True se o ponto em NDC estiver dentro da caixa."""
        return self.min_x() <= px <= self.max_x() and self.min_y() <= py <= self.max_y()


class CollisionPickingDemo:
    """Gerencia simulação, picking e renderização de caixas em colisão."""

    def __init__(self) -> None:
        self.random = Random(42)
        self.entities: list[BoxEntity] = []
        self.colliding_pairs: set[tuple[int, int]] = set()
        self._build_entities()

    def _build_entities(self) -> None:
        self.entities.clear()
        palette = [(1.0, 0.4, 0.4), (0.4, 1.0, 0.5), (0.4, 0.7, 1.0), (1.0, 0.9, 0.4), (0.9, 0.5, 1.0)]
        for i in range(6):
            self.entities.append(
                BoxEntity(
                    x=self.random.uniform(-0.8, 0.8),
                    y=self.random.uniform(-0.6, 0.6),
                    half_size=self.random.uniform(0.07, 0.11),
                    vx=self.random.uniform(-0.006, 0.006),
                    vy=self.random.uniform(-0.006, 0.006),
                    color=palette[i % len(palette)],
                )
            )

    def _aabb_overlap(self, a: BoxEntity, b: BoxEntity) -> bool:
        return not (
            a.max_x() < b.min_x()
            or a.min_x() > b.max_x()
            or a.max_y() < b.min_y()
            or a.min_y() > b.max_y()
        )

    def _solve_wall_bounce(self, entity: BoxEntity) -> None:
        if entity.min_x() < -1.0 or entity.max_x() > 1.0:
            entity.vx *= -1.0
        if entity.min_y() < -1.0 or entity.max_y() > 1.0:
            entity.vy *= -1.0

        entity.x = max(-1.0 + entity.half_size, min(1.0 - entity.half_size, entity.x))
        entity.y = max(-1.0 + entity.half_size, min(1.0 - entity.half_size, entity.y))

    def update(self) -> None:
        """Atualiza posições, resolve paredes e detecta colisões AABB."""
        for entity in self.entities:
            entity.x += entity.vx
            entity.y += entity.vy
            self._solve_wall_bounce(entity)

        self.colliding_pairs.clear()
        n = len(self.entities)
        for i in range(n):
            for j in range(i + 1, n):
                if self._aabb_overlap(self.entities[i], self.entities[j]):
                    self.colliding_pairs.add((i, j))

    def pick_at(self, px: float, py: float) -> None:
        """Seleciona entidade clicada; se nenhuma, limpa seleção."""
        selected_any = False
        for entity in self.entities:
            entity.selected = False

        for entity in reversed(self.entities):
            if entity.contains(px, py):
                entity.selected = True
                selected_any = True
                break

        if not selected_any:
            for entity in self.entities:
                entity.selected = False

    def _draw_text(self, x: float, y: float, text: str) -> None:
        glColor3f(0.9, 0.9, 0.95)
        glRasterPos2f(x, y)
        for char in text:
            glutBitmapCharacter(GLUT_BITMAP_HELVETICA_12, ord(char))

    def _draw_box(self, entity: BoxEntity, colliding: bool) -> None:
        base = entity.color
        if colliding:
            glColor3f(1.0, 0.2, 0.2)
        else:
            glColor3f(*base)

        glBegin(GL_QUADS)
        glVertex2f(entity.min_x(), entity.min_y())
        glVertex2f(entity.max_x(), entity.min_y())
        glVertex2f(entity.max_x(), entity.max_y())
        glVertex2f(entity.min_x(), entity.max_y())
        glEnd()

        if entity.selected:
            glColor3f(1.0, 1.0, 1.0)
            glBegin(GL_LINE_LOOP)
            glVertex2f(entity.min_x() - 0.01, entity.min_y() - 0.01)
            glVertex2f(entity.max_x() + 0.01, entity.min_y() - 0.01)
            glVertex2f(entity.max_x() + 0.01, entity.max_y() + 0.01)
            glVertex2f(entity.min_x() - 0.01, entity.max_y() + 0.01)
            glEnd()

    def render(self) -> None:
        glClear(GL_COLOR_BUFFER_BIT)
        glMatrixMode(GL_MODELVIEW)
        glLoadIdentity()

        colliding_indices = {idx for pair in self.colliding_pairs for idx in pair}
        for i, entity in enumerate(self.entities):
            self._draw_box(entity, i in colliding_indices)

        self._draw_text(-0.96, -0.94, f"colisoes={len(self.colliding_pairs)} | clique para selecionar | R reset")
        glutSwapBuffers()


demo = CollisionPickingDemo()


def configure_close_behavior() -> None:
    try:
        glutSetOption(GLUT_ACTION_ON_WINDOW_CLOSE, GLUT_ACTION_GLUTMAINLOOP_RETURNS)
    except Exception:
        pass


def display() -> None:
    demo.render()


def keyboard(key: bytes, _: int, __: int) -> None:
    if key in (b"r", b"R"):
        demo._build_entities()
    elif key == b"\x1b":
        raise SystemExit("ESC")


def mouse(button: int, state: int, x: int, y: int) -> None:
    if button != GLUT_LEFT_BUTTON or state != GLUT_DOWN:
        return

    width = glutGet(GLUT_WINDOW_WIDTH)
    height = glutGet(GLUT_WINDOW_HEIGHT)
    px = 2.0 * (x / width) - 1.0
    py = 1.0 - 2.0 * (y / height)
    demo.pick_at(px, py)
    glutPostRedisplay()


def tick(_: int) -> None:
    demo.update()
    glutPostRedisplay()
    glutTimerFunc(16, tick, 0)


def main() -> None:
    glutInit()
    configure_close_behavior()
    glutInitDisplayMode(GLUT_DOUBLE | GLUT_RGB)
    glutInitWindowSize(1200, 850)
    glutCreateWindow(b"Aula 09 - Picking e Colisoes AABB")
    glClearColor(0.04, 0.04, 0.07, 1.0)

    glutDisplayFunc(display)
    glutKeyboardFunc(keyboard)
    glutMouseFunc(mouse)
    glutTimerFunc(0, tick, 0)

    print("Controles: clique, R, ESC")
    glutMainLoop()


if __name__ == "__main__":
    main()

File: glut_opengl_estudos/09_picking_colisoes_aabb.cpp
/*
Aula 09 — Picking e colisões AABB (C++)

Objetivo pedagógico:
- Implementar detecção de clique sobre objetos (picking 2D).
- Implementar colisão Axis-Aligned Bounding Box (AABB).
- Explicar claramente conversão de coordenadas de tela para mundo.
*/

#include <GL/glut.h>
#include <string>
#include <vector>
#include "cpp_glut_utils.hpp"

namespace {

struct Box {
    float x;
    float y;
    float w;
    float h;
    float r;
    float g;
    float b;
    bool selected = false;
};

std::vector<Box> boxes = {
    {20.0f, 65.0f, 16.0f, 12.0f, 0.9f, 0.4f, 0.4f, false},
    {52.0f, 52.0f, 20.0f, 14.0f, 0.4f, 0.8f, 1.0f, false},
    {78.0f, 30.0f, 14.0f, 18.0f, 0.5f, 1.0f, 0.6f, false}
};

int selectedIndex = -1;

void drawRect(const Box& box) {
    const float hw = box.w * 0.5f;
    const float hh = box.h * 0.5f;

    if (box.selected) {
        // Objeto selecionado fica mais claro para feedback visual.
        glColor3f(aula::clamp(box.r + 0.25f, 0.0f, 1.0f), aula::clamp(box.g + 0.25f, 0.0f, 1.0f), aula::clamp(box.b + 0.25f, 0.0f, 1.0f));
    } else {
        glColor3f(box.r, box.g, box.b);
    }

    glBegin(GL_QUADS);
    glVertex2f(box.x - hw, box.y - hh);
    glVertex2f(box.x + hw, box.y - hh);
    glVertex2f(box.x + hw, box.y + hh);
    glVertex2f(box.x - hw, box.y + hh);
    glEnd();
}

bool pointInAabb(float px, float py, const Box& b) {
    const float left = b.x - b.w * 0.5f;
    const float right = b.x + b.w * 0.5f;
    const float bottom = b.y - b.h * 0.5f;
    const float top = b.y + b.h * 0.5f;
    return px >= left && px <= right && py >= bottom && py <= top;
}

bool intersects(const Box& a, const Box& b) {
    const float aLeft = a.x - a.w * 0.5f;
    const float aRight = a.x + a.w * 0.5f;
    const float aBottom = a.y - a.h * 0.5f;
    const float aTop = a.y + a.h * 0.5f;

    const float bLeft = b.x - b.w * 0.5f;
    const float bRight = b.x + b.w * 0.5f;
    const float bBottom = b.y - b.h * 0.5f;
    const float bTop = b.y + b.h * 0.5f;

    return !(aRight < bLeft || aLeft > bRight || aTop < bBottom || aBottom > bTop);
}

void display() {
    glClear(GL_COLOR_BUFFER_BIT);
    glMatrixMode(GL_MODELVIEW);
    glLoadIdentity();

    for (const auto& b : boxes) drawRect(b);

    // Mostra colisão entre primeira e segunda caixas.
    const bool collision = intersects(boxes[0], boxes[1]);
    glColor3f(collision ? 1.0f : 0.8f, collision ? 0.5f : 0.9f, collision ? 0.5f : 1.0f);
    aula::drawBitmapText(2.0f, 96.0f, collision ? "AABB: COLISAO" : "AABB: sem colisao");

    glutSwapBuffers();
}

void reshape(int width, int height) {
    if (height == 0) height = 1;
    glViewport(0, 0, width, height);
    glMatrixMode(GL_PROJECTION);
    glLoadIdentity();
    glOrtho(0.0, 100.0, 0.0, 100.0, -1.0, 1.0);
}

void special(int key, int, int) {
    if (selectedIndex < 0) return;

    constexpr float step = 1.8f;
    if (key == GLUT_KEY_LEFT) boxes[selectedIndex].x -= step;
    else if (key == GLUT_KEY_RIGHT) boxes[selectedIndex].x += step;
    else if (key == GLUT_KEY_UP) boxes[selectedIndex].y += step;
    else if (key == GLUT_KEY_DOWN) boxes[selectedIndex].y -= step;

    glutPostRedisplay();
}

void keyboard(unsigned char key, int, int) {
    if (key == 27) std::exit(0);
}

void mouse(int button, int state, int x, int y) {
    if (button != GLUT_LEFT_BUTTON || state != GLUT_DOWN) return;

    const int width = glutGet(GLUT_WINDOW_WIDTH);
    const int height = glutGet(GLUT_WINDOW_HEIGHT);

    // Conversão de pixel para mundo ortográfico [0,100]x[0,100].
    const float worldX = 100.0f * static_cast<float>(x) / static_cast<float>(width);
    const float worldY = 100.0f * (1.0f - static_cast<float>(y) / static_cast<float>(height));

    selectedIndex = -1;
    for (std::size_t i = 0; i < boxes.size(); ++i) {
        boxes[i].selected = pointInAabb(worldX, worldY, boxes[i]);
        if (boxes[i].selected) selectedIndex = static_cast<int>(i);
    }

    glutPostRedisplay();
}

} // namespace

int main(int argc, char** argv) {
    glutInit(&argc, argv);
    aula::configureCloseBehavior();
    glutInitDisplayMode(GLUT_DOUBLE | GLUT_RGB);
    glutInitWindowSize(1050, 760);
    glutCreateWindow("Aula 09 - Picking e AABB (C++)");

    glClearColor(0.04f, 0.04f, 0.07f, 1.0f);
    glutDisplayFunc(display);
    glutReshapeFunc(reshape);
    glutKeyboardFunc(keyboard);
    glutSpecialFunc(special);
    glutMouseFunc(mouse);
    glutMainLoop();
    return 0;
}
